src/lidar_tools/dsm_functions.py

```diff
@@ -223,106 +339,342 @@ def create_dem_stage(dem_filename='dem_output.tif', pointcloud_resolution=1.,
+        cop_da = cop_da.rio.reproject_match(match_grid_da,resampling=rasterio.enums.Resampling.cubic)
+    return cop_da
+
+
+
+def confirm_3dep_vertical(raster_fn: str,
+                bare_diff_tolerance: float = 3.0) -> bool:
+    lidar_da = rioxarray.open_rasterio(raster_fn,masked=True).squeeze()
+    worldcover_da = fetch_worldcover(raster_fn,lidar_da)
+    cop30_da = fetch_cop30(raster_fn,lidar_da)
+    lidar_da_masked,worldcover_da_masked,cop30_da_masked = common_mask([lidar_da,worldcover_da,cop30_da],apply=True)
+    dem_diff = lidar_da_masked - cop30_da_masked
+    ## Mask out bare and sparse vegetation class
+    bare_sparse_mask = worldcover_da_masked == 60
+    dem_diff_bare = dem_diff.where(bare_sparse_mask,np.nan)
+    median_diff = np.nanmedian(dem_diff_bare.values)
+    print(f"Observed difference between COP30 EGM2008 and 3DEP LiDAR DSM over bareground and sparse vegetation surfaces is {median_diff:.2f} m")
+    if np.abs(median_diff) <= bare_diff_tolerance:
+        #this means that both COP30 and 3DEP LiDAR DSM are with respect to geoid
+        print("Looks like the 3DEP height estimates are with respect to geiod, will apply vertical datum shift to return heights with respect to ellipsoid")
+        out = True
+    else:
+        #this means that 3DEP LiDAR DSM is with respect to ellipsoid
+        print("Looks like the 3DEP height estimates are already with respect to ellipsoid, geoid to ellipsoid transformation will not be attempted")
+        out = False
+    return out
 
+def check_raster_validity(raster_fn: str) -> bool:
+    """
+    Check if a raster file is valid and can be opened using rioxarray and CRS check
 
+    Parameters
+    ----------
+    raster_fn : str
+        Path to the raster file.
 
+    Returns
+    -------
+    bool
+        True if the raster file is valid, False otherwise.
+    """
+    da = rioxarray.open_rasterio(raster_fn,masked=True).squeeze()
+    if da.rio.crs is None:
+        #print(f"Raster {raster_fn} does not have a valid CRS.")
+        out = False
+    else:
+        #print(f"Raster {raster_fn} has a valid CRS.")
+        out = True
+    return out   
+def gdal_warp(src_fn: str,
+                dst_fn: str, 
+                src_srs: str, 
+                dst_srs: str, 
+                res: float = 1.0,
+                resampling_alogrithm: str ='cubic') -> None:
+    tolerance = 0
+    resampling_mapping = {"nearest":  gdalconst.GRA_NearestNeighbour, "bilinear": gdalconst.GRA_Bilinear,
+                  "cubic": gdalconst.GRA_Cubic, "cubic_spline": gdalconst.GRA_CubicSpline}
+    resampling_alg = resampling_mapping[resampling_alogrithm]
+    ds = gdal.Warp(dst_fn, src_fn,
+                   resampleAlg=resampling_alg,
+                   srcSRS=src_srs, xRes=res, yRes=res,
+                   dstSRS=dst_srs, errorThreshold=tolerance,
+                   targetAlignedPixels=True,
+                   creationOptions=['COMPRESS=LZW','TILED=YES','COPY_SRC_OVERVIEWS=YES'],
```

Comment:
```suggestion
                   format='COG'
```
I haven't used GDAL python bindings very much, but I wonder if that wouldn't be simpler? Would also automatically add overviews

Reply:
Yes, COG is better option, as long as we can specify `gauss` resampling option to create overviews

Reply:
Good point Scott, COG would be better. 
As per David's note, I checked the available overview resampling methods for the COG driver (https://gdal.org/en/latest/drivers/raster/cog.html), and Gaussian resampling is not available. I am leaving this as is, and added a note to revisit this in the future!